kyma-incubator/service-catalog
# Replace default service plan with webhook

This PR is a 

**What this PR does / why we need it**:

This PR introduces the `defaultServicePlan` as a part of ServiceInstance's mutation webhook.

`defaultServicePlan` feature adds the ServicePlan reference to the ServiceInstance if it wasn't provided and if there is only one ServicePlan which refers to the ServiceClass from which we provision the ServiceInstance.

**How to test**
1. Create our SC on webhooks setup
2. `svcat get classes` - should return our 3 default classes
3. Create service instance from the following yaml:
```
apiVersion: servicecatalog.k8s.io/v1beta1
kind: ServiceInstance
metadata:
  name: si
spec:
  clusterServiceClassExternalName: azure-service-broker
```
- kubectl create -f abovefile.yaml -n qa

4. Created ServiceInstance should now be filled with plan reference:
```
   "spec": {
                "clusterServiceClassExternalName": "azure-service-broker",
                "clusterServiceClassRef": {
                    "name": "84dad23a-794d-404f-92a1-5e73e91a74af"
                },
                "clusterServicePlanExternalName": "default",
                "clusterServicePlanRef": {
                    "name": "2e02da10-c058-4044-b02d-4c58f70b2104"
                },
```
5. Update ServiceInstance - delete all references about plans and save.
6. Plan references should be automatically restored - should be the same as in the step 4.

7. Now create SI from this file:
```
apiVersion: servicecatalog.k8s.io/v1beta1
kind: ServiceInstance
metadata:
  name: si
spec:
  clusterServiceClassExternalName: redis
```
8. You should see the following error:
<img width="1590" alt="Screenshot 2019-03-28 at 14 30 15" src="https://user-images.githubusercontent.com/21159554/55163559-4c4cde00-516a-11e9-99fd-4967b351fa6c.png">
